module/zfs/bpobj.c

```diff
@@ -242,119 +279,167 @@ bpobj_iterate_impl(bpobj_t *bpo, bpobj_itor_t func, void *arg, dmu_tx_t *tx,
 	}
-	if (free) {
-		VERIFY3U(0, ==, dmu_free_range(bpo->bpo_os, bpo->bpo_object,
-		    (i + 1) * sizeof (blkptr_t), DMU_OBJECT_END, tx));
-	}
-	if (err || !bpo->bpo_havesubobj || bpo->bpo_phys->bpo_subobjs == 0)
-		goto out;
+	return (err);
+}
 
-	ASSERT(bpo->bpo_havecomp);
-	err = dmu_object_info(bpo->bpo_os, bpo->bpo_phys->bpo_subobjs, &doi);
-	if (err) {
-		mutex_exit(&bpo->bpo_lock);
-		return (err);
+static void
+bpobj_iterate_cleanup(list_t *stack, int err)
+{
+	bpobj_t *bpo;
+	bpobj_info_t *bpi;
+
+	while ((bpi = list_remove_head(stack)) != NULL) {
+		ASSERT3U(err, !=, 0);
+		bpo = bpi->bpi_bpo;
+		ASSERT3P(bpo, !=, NULL);
+		if (bpi->bpi_parent != NULL) {
+			bpobj_close(bpo);
+			kmem_free(bpi->bpi_bpo, sizeof (bpobj_t));
+		}
+		kmem_free(bpi, sizeof (bpobj_info_t));
 	}
-	ASSERT3U(doi.doi_type, ==, DMU_OT_BPOBJ_SUBOBJ);
-	epb = doi.doi_data_block_size / sizeof (uint64_t);
-
-	for (i = bpo->bpo_phys->bpo_num_subobjs - 1; i >= 0; i--) {
-		uint64_t *objarray;
-		uint64_t offset, blkoff;
-		bpobj_t sublist;
-		uint64_t used_before, comp_before, uncomp_before;
-		uint64_t used_after, comp_after, uncomp_after;
+}
 
-		offset = i * sizeof (uint64_t);
-		blkoff = P2PHASE(i, epb);
+static int
+bpobj_iterate_impl(bpobj_t *initial_bpo, bpobj_itor_t func, void *arg,
+    dmu_tx_t *tx, boolean_t free)
+{
+	list_t stack;
+	int err = 0;
 
-		if (dbuf == NULL || dbuf->db_offset > offset) {
-			if (dbuf)
-				dmu_buf_rele(dbuf, FTAG);
-			err = dmu_buf_hold(bpo->bpo_os,
-			    bpo->bpo_phys->bpo_subobjs, offset, FTAG, &dbuf, 0);
-			if (err)
-				break;
+	list_create(&stack, sizeof (bpobj_info_t),
+	    offsetof(bpobj_info_t, bpi_node));
+	for (bpobj_info_t *bpi = bpi_alloc(initial_bpo, NULL, 0, 0);
+	    bpi != NULL; bpi = list_remove_head(&stack)) {
+		bpobj_t *bpo = bpi->bpi_bpo;
+		mutex_enter(&bpo->bpo_lock);
+		ASSERT(bpobj_is_open(bpo));
+		if (free)
+			dmu_buf_will_dirty(bpo->bpo_dbuf, tx);
+
+		err = bpobj_iterate_blkptrs(bpi, func, arg, tx, free);
+		if (err != 0) {
+			mutex_exit(&bpo->bpo_lock);
+			/* push back onto stack for cleanup function */
+			list_insert_head(&stack, bpi);
+			goto out;
 		}
-
-		ASSERT3U(offset, >=, dbuf->db_offset);
-		ASSERT3U(offset, <, dbuf->db_offset + dbuf->db_size);
-
-		objarray = dbuf->db_data;
-		err = bpobj_open(&sublist, bpo->bpo_os, objarray[blkoff]);
-		if (err)
-			break;
-		if (free) {
-			err = bpobj_space(&sublist,
-			    &used_before, &comp_before, &uncomp_before);
-			if (err != 0) {
-				bpobj_close(&sublist);
-				break;
+		/* A subobj with no subobjs below it can be cleaned up. */
+		if (bpo->bpo_phys->bpo_num_subobjs == 0 ||
+		    !bpo->bpo_havesubobj ||
+		    bpo->bpo_phys->bpo_subobjs == 0) {
+			/*
+			 * If there are no entries, there should
+			 * be no bytes.
+			 */
+			if (bpobj_is_empty(bpo)) {
+				ASSERT0(bpo->bpo_phys->bpo_bytes);
+				ASSERT0(bpo->bpo_phys->bpo_comp);
+				ASSERT0(bpo->bpo_phys->bpo_uncomp);
 			}
+			if (free && (bpi->bpi_parent != NULL)) {
```

Comment:
Would it be clearer to have (bpi->bpi_bpo != initial_bpo) ?  Because it's basically a proxy for that.  We don't free it, and we don't try to reference it's "parent" (i.e. to decrement bpo_num_subobjs) because it doesn't have one.

Reply:
Yeah I was overly focused on "free == TRUE" case (since that was my test) and that shows in the comments.  I'll use "process" as the verb instead of "freed" and "cleaned up".

Reply:
yeah I think `bpi->bpi_bpo != initial_bpo` might be more clear.  That way the reader might realize (even without a comment) that the point is to not free the passed-in bpo.